### PR TITLE
Implement Pico

### PR DESCRIPTION
**Pull request by alexdovzhanyn**
Pico is a new zero-knowledge peer to peer messaging protocol based on TCP that we'll be using as the networking layer of the network. This pull request pulls in the [Pico](https://github.com/ElixiumNetwork/pico) library that we've put together. This library offers numerous improvements over our current networking implementation, namely:
* A DSL for message routing
* Parallelization of message processing (we no longer have one router process that queues messages)
* Improved message parsing (both logic and speed)

*This pull request must be tested while pointing to the implement-pico branch of elixium_core*

### CODE REVIEW

File: lib/peer/router.ex [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+      {:missing_blocks, fork_chain} ->
+        # We've discovered a fork, but we can't rebuild the fork chain without
+        # some blocks. Let's request them from our peer.
+        Pico.message(conn, "BLOCK_QUERY_REQUEST", %{index: :binary.decode_unsigned(hd(fork_chain).index) - 1})
+
+      :invalid ->
+        Logger.info("Recieved invalid block at index #{:binary.decode_unsigned(block.index)}.")
+
+      :ignore -> :ignore
```
Comment (fantypants) [inline lib/peer/router.ex:37]: might need to have the edge case ```_->``` just in case we get a random exception

File: lib/peer/router.ex [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+  message "BLOCK_QUERY_RESPONSE", response do
+    orphans_ahead =
+      Ledger.last_block().index
+      |> :binary.decode_unsigned()
+      |> Kernel.+(1)
+      |> Orphan.blocks_at_height()
+      |> length()
+
+    if orphans_ahead > 0 do
```
Comment (fantypants) [inline lib/peer/router.ex:53]: We should start highlighting this stuff to be removed, if we dont use it within a certain time frame

File: lib/peer/router.ex [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+  message "NEW_OUTBOUND_CONNECTION", _ do
+    # Let's ask our peer for new blocks, if there
+    # are any. We'll ask for all blocks starting from our current index minus
+    # 120 (4 hours worth of blocks before we disconnected) just in case there
+    # was a fork after we disconnected.
+
+    starting_at =
+      case Ledger.last_block() do
+        :err -> 0
```
Comment (fantypants) [inline lib/peer/router.ex:157]: same as above, edge case if needed, or if it's only two cases like before it may be cleaner to use a  ```with-else```

**Review by fantypants — APPROVED** [review]
Cleaner, way more concise, this is looking really really clean now.

File: lib/peer/router.ex (hunk 1 above)
Comment (alexdovzhanyn) [reply lib/peer/router.ex:37]: These values are passed directly from functions we control -- we'll never get anything other than these values as a response here

File: lib/peer/router.ex (hunk 2 above)
Comment (alexdovzhanyn) [reply lib/peer/router.ex:53]: This is a TODO, left it here so we don't forget to implement it

File: lib/peer/router.ex (hunk 3 above)
Comment (alexdovzhanyn) [reply lib/peer/router.ex:157]: With is usually used when chaining results together -- we're not doing that here so I think a case statement is the cleanest way to handle this

**Pull request merged**